### DISCUSSION TITLE
useFormContext returns a new object on each call, is this expected?

### DISCUSSION THREAD

**Discussion by RebeccaStevens** [react-hook-form/react-hook-form#2804] [Q&A] [answered]
I'm using the eslint rule "react-hooks/exhaustive-deps" so I need to pass the value returned from `useFormContext` as a dependency to hooks such as `useEffect` if they need to use one or more of the form methods. But this is causing an infinite re-render as the object returned from `useFormContext` is different each time.

For example: 
```tsx
function MyComponent() {
  const form = useFormContext();
  const value = form.watch('foo');

  useEffect(() => {
    form.setValue('foo', 'bar');
  }, [form]);

  return (<div>{value}</div>);
}
```

I believe in this example, if I deconstructed `useFormContext` and simply passed `setValue` as the dependency (or just `form.setValue`), everything would be ok. However, in my actual case this isn't really an option as I need to then pass the entire `form` value to a 3rd-party function that then performs some action.

I have figured out that I can use `useRef` like this to get around the issue:

```ts
const form = useRef(useFormContext()).current;
```

Would this be considered a safe way of doing things or would you recommend another option? Or should `useFormContext` just be returning the same object each time?

**Comment by bluebill1049** [comment, marked as answer]
consider to use form.setValue at the dep.

**Reply by RebeccaStevens** [reply]
Yes, I considered that but I need to pass everything returned from `useFormContext` to a 3rd-partly function inside the `useEffect` as that's what that functions wants.

**Reply by RebeccaStevens** [reply]
Do I need to get that 3rd-partly function changed?

**Reply by bluebill1049** [reply]
we don't memo the whole object which returns from `useForm`, you should put the individual methods in the dep.

**Reply by justin-hoops** [reply]
You should though, otherwise it rerenders every time. Took me a while to nail react hook forms as the culprit of my rerenders.
